libr/debug/debug.c

```diff
@@ -516,70 +516,99 @@ R_API bool r_debug_set_arch(RDebug *dbg, const char *arch, int bits) {
  *
  * TODO: Add support for reverse stack architectures
  */
-R_API ut64 r_debug_execute(RDebug *dbg, const ut8 *buf, int len, bool restore, bool ignore_stack) {
-	int orig_sz;
+R_API bool r_debug_execute(RDebug *dbg, const ut8 *buf, int len, R_OUT ut64 *ret, bool restore, bool ignore_stack) {
+	RCore *core;
 	ut8 stack_backup[4096];
 	ut8 *pc_backup = NULL, *reg_backup = NULL;
-	RRegItem *ri, *risp, *ripc;
-	ut64 rsp, rpc, ra0 = 0LL;
+	int reg_backup_sz;
+	RRegItem *ri_sp, *ri_pc, *ri_ret;
+	ut64 reg_sp, reg_pc, bp_addr;
+
+	r_return_val_if_fail (dbg && buf && len > 0, false);
+	core = dbg->corebind.core;
```

Comment:
the idea behind corebind is that the RCore structure is opaque. so that is should be solved that way. why is the current seek changing from core? rdebug shouldnt be the one doing this if you restore the PC register after that..

Also. you can add an RCoreSeek() caallback in the RCoreBind structure and do `dbg->corebind.seek()` instead of accessing the offset field directly. That would be more legal.. but i dont think debug should be changing the core.offset

Reply:
> rdebug shouldnt be the one doing this if you restore the PC register after that

The problem is that the PC is restored, but seek is not. I believe that seek is getting forced forward when a breakpoint is hit, which seems incorrect unless we simply accept it and then move it in the command function. But it's also used elsewhere such as in native debug code, which would also result in seek being moved unexpectedly.

> but i dont think debug should be changing the core.offset

That's what I'm trying to fix. Maybe we should be able to just not move seek when a breakpoint is hit but that would probably require a lot more effort.

Reply:
then you may want to restore the seek in RCore afer calling this function. Core is the only one who should be updating the core->offset variable

Reply:
We can just do it in cmd_debug() then, and note the seek behavior.